### DISCUSSION TITLE
Is there  any method to source some python code when press `!` ?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [inducer/pudb#580] [Q&A] [answered]
I hope every time in pudb, I press `!` to enter the REPL will source some python code like `from matplotlib import pyplot as plt; ...` automatically. Is there any method to do it? TIA!

**Comment by inducer** [comment]
There's no such option right now, but I think that'd be relatively straightforward to add. I think the functionality makes sense, perhaps configurable in settings, and I'd be happy to consider a PR.

**Comment by asmeurer** [comment, marked as answer]
This can be done with a custom shell https://documen.tician.de/pudb/shells.html#custom-shells. Just exec the code you want into the `_globals` then pass it through to the existing [shell function](https://github.com/inducer/pudb/blob/main/pudb/shell.py). For example, if you use the IPython shell:

```py
from pudb.shell import run_ipython_shell

def custom_shell(_globals, _locals):
    exec('from matplotlib import pyplot as plt', _globals, _locals)
    return run_ipython_shell(_globals, _locals)
```

**Reply by cohml** [reply]
The same could also probably be accomplished in a more "Pythonic" fashion using a decorator:

```python
from pudb.shell import run_ipython_shell

def custom_code_sourcer(func):
    def hook(*args, **kwargs):
        # put all your extra code to run here;
        # e.g., from matplotlib import pyplot as plt
        return func(*args, **kwargs)
    return hook

@custom_code_sourcer
def custom_shell(_globals, _locals):
    return run_ipython_shell(_globals, _locals)
```
